File: cursorpaste.m
#import <Cocoa/Cocoa.h>
#import <ApplicationServices/ApplicationServices.h>

static const CGFloat kMaxSide = 220.0;
static const CGFloat kMinSide = 40.0;
static const CGFloat kHoverOffsetX = 14.0;

static void LogHelper(NSString *message) {
    NSFileHandle *stderrHandle = [NSFileHandle fileHandleWithStandardError];
    NSString *line = [message stringByAppendingString:@"\n"];
    [stderrHandle writeData:[line dataUsingEncoding:NSUTF8StringEncoding]];
}

@interface CursorPasteController : NSObject
@property(nonatomic, strong) NSImage *image;
@property(nonatomic, strong) NSURL *fileURL;
@property(nonatomic, assign) BOOL trusted;
@property(nonatomic, assign) NSSize thumbSize;

@property(nonatomic, strong) NSPanel *hoverPanel;
@property(nonatomic, strong) NSImageView *hoverImageView;
@property(nonatomic, strong) NSMutableArray *monitors;
@property(nonatomic, strong) NSTimer *followTimer;

@property(nonatomic, assign) BOOL finished;
@property(nonatomic, assign) BOOL pasteInFlight;
- (void)sendLeftClickAt:(NSPoint)point;
- (void)handleGlobalMouseUp:(NSEvent *)event;
@end

@implementation CursorPasteController

- (instancetype)initWithImage:(NSImage *)image fileURL:(NSURL *)fileURL trusted:(BOOL)trusted {
    self = [super init];
    if (!self) return nil;

    _image = image;
    _fileURL = fileURL;
    _trusted = trusted;
    _monitors = [NSMutableArray array];

    NSSize imageSize = image.size;
    CGFloat longestSide = MAX(imageSize.width, imageSize.height);
    CGFloat scale = longestSide > 0 ? MIN(1.0, kMaxSide / longestSide) : 1.0;
    _thumbSize = NSMakeSize(MAX(kMinSide, imageSize.width * scale),
                            MAX(kMinSide, imageSize.height * scale));

    _hoverPanel = [[NSPanel alloc] initWithContentRect:NSMakeRect(0, 0, _thumbSize.width, _thumbSize.height)
                                             styleMask:NSWindowStyleMaskBorderless | NSWindowStyleMaskNonactivatingPanel
                                               backing:NSBackingStoreBuffered
                                                 defer:NO];
    _hoverPanel.floatingPanel = YES;
    _hoverPanel.level = NSScreenSaverWindowLevel;
    _hoverPanel.backgroundColor = [NSColor clearColor];
    _hoverPanel.opaque = NO;
    _hoverPanel.hasShadow = YES;
    _hoverPanel.ignoresMouseEvents = YES;
    _hoverPanel.collectionBehavior = NSWindowCollectionBehaviorCanJoinAllSpaces |
                                     NSWindowCollectionBehaviorFullScreenAuxiliary |
                                     NSWindowCollectionBehaviorStationary;
    _hoverPanel.alphaValue = 0.92;

    _hoverImageView = [[NSImageView alloc] initWithFrame:NSMakeRect(0, 0, _thumbSize.width, _thumbSize.height)];
    _hoverImageView.image = image;
    _hoverImageView.imageScaling = NSImageScaleProportionallyUpOrDown;
    _hoverImageView.wantsLayer = YES;
    _hoverImageView.layer.cornerRadius = 8.0;
    _hoverImageView.layer.masksToBounds = YES;
    _hoverImageView.layer.borderWidth = 2.0;
    _hoverImageView.layer.borderColor = [[NSColor colorWithWhite:1.0 alpha:0.85] CGColor];
    _hoverPanel.contentView = _hoverImageView;

    return self;
}

- (void)start {
    [self showHover];
    [self repositionHover];

    self.followTimer = [NSTimer scheduledTimerWithTimeInterval:(1.0 / 120.0)
                                                        target:self
                                                      selector:@selector(repositionHover)
                                                      userInfo:nil
                                                       repeats:YES];
    [[NSRunLoop mainRunLoop] addTimer:self.followTimer forMode:NSRunLoopCommonModes];

    __weak typeof(self) weakSelf = self;
    id monitor = [NSEvent addGlobalMonitorForEventsMatchingMask:NSEventMaskKeyDown
                                                        handler:^(NSEvent *event) {
        if (event.keyCode == 53) {
            [weakSelf finish:1];
        }
    }];
    if (monitor) {
        [self.monitors addObject:monitor];
    }
    id mouseMonitor = [NSEvent addGlobalMonitorForEventsMatchingMask:NSEventMaskLeftMouseUp
                                                             handler:^(NSEvent *event) {
        [weakSelf handleGlobalMouseUp:event];
    }];
    if (mouseMonitor) {
        [self.monitors addObject:mouseMonitor];
    }
}

- (void)handleGlobalMouseUp:(NSEvent *)event {
    if (self.pasteInFlight || self.finished) {
        return;
    }
    self.pasteInFlight = YES;
    [self performPasteFallbackAt:[NSEvent mouseLocation]];
}

- (void)performPasteFallbackAt:(NSPoint)screenPoint {
    [self copyToPasteboard];
    LogHelper(@"cursorpaste: copied screenshot to pasteboard");

    if (!self.trusted) {
        LogHelper(@"cursorpaste: click-to-paste blocked because Accessibility trust is missing");
        [self finish:1];
        return;
    }

    [self hideHover];

    LogHelper([NSString stringWithFormat:@"cursorpaste: sending focus click at %.1f, %.1f", screenPoint.x, screenPoint.y]);
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.05 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self sendLeftClickAt:screenPoint];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.22 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            NSRunningApplication *frontmost = [[NSWorkspace sharedWorkspace] frontmostApplication];
            NSString *bundleID = frontmost.bundleIdentifier ?: @"";
            if ([self shouldTypeFilePathForBundleID:bundleID]) {
                LogHelper([NSString stringWithFormat:@"cursorpaste: frontmost app %@ looks like a terminal; typing file path", bundleID]);
                [self sendFilePathText];
            } else {
                LogHelper([NSString stringWithFormat:@"cursorpaste: frontmost app %@ supports normal paste; sending Cmd+V", bundleID]);
                [self sendPaste];
            }
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.15 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                [self finish:0];
            });
        });
    });
}

- (void)copyToPasteboard {
    NSPasteboard *pb = [NSPasteboard generalPasteboard];
    [pb clearContents];

    NSBitmapImageRep *rep = nil;
    NSData *tiff = [self.image TIFFRepresentation];
    NSData *png = nil;
    if (tiff) {
        rep = [NSBitmapImageRep imageRepWithData:tiff];
        png = [rep representationUsingType:NSBitmapImageFileTypePNG properties:@{}];
    }

    NSPasteboardItem *item = [[NSPasteboardItem alloc] init];
    if (png) {
        [item setData:png forType:NSPasteboardTypePNG];
    }
    if (tiff) {
        [item setData:tiff forType:NSPasteboardTypeTIFF];
    }

    NSMutableArray *objects = [NSMutableArray arrayWithObject:self.fileURL];
    if (png || tiff) {
        [objects insertObject:item atIndex:0];
    }
    [pb writeObjects:objects];
}

- (void)sendPaste {
    CGEventSourceRef src = CGEventSourceCreate(kCGEventSourceStateCombinedSessionState);
    if (!src) {
        LogHelper(@"cursorpaste: failed to create keyboard event source");
        return;
    }

    CGKeyCode vKey = 0x09;
    CGEventRef down = CGEventCreateKeyboardEvent(src, vKey, true);
    CGEventRef up = CGEventCreateKeyboardEvent(src, vKey, false);
    if (down) {
        CGEventSetFlags(down, kCGEventFlagMaskCommand);
        CGEventPost(kCGHIDEventTap, down);
        CFRelease(down);
    }
    if (up) {
        CGEventSetFlags(up, kCGEventFlagMaskCommand);
        CGEventPost(kCGHIDEventTap, up);
        CFRelease(up);
    }
    CFRelease(src);
}

- (BOOL)shouldTypeFilePathForBundleID:(NSString *)bundleID {
    static NSSet<NSString *> *terminalBundleIDs;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        terminalBundleIDs = [NSSet setWithArray:@[
            @"com.apple.Terminal",
            @"com.googlecode.iterm2",
            @"dev.warp.Warp-Stable",
            @"dev.warp.Warp",
            @"com.mitchellh.ghostty",
            @"org.alacritty",
            @"net.kovidgoyal.kitty",
            @"co.zeit.hyper",
            @"org.wezfurlong.wezterm"
        ]];
    });
    return [terminalBundleIDs containsObject:bundleID];
}

- (NSString *)shellEscapedPath {
    NSString *path = self.fileURL.path ?: @"";
    NSString *escaped = [path stringByReplacingOccurrencesOfString:@"'" withString:@"'\"'\"'"];
    return [NSString stringWithFormat:@"'%@'", escaped];
}

- (void)sendFilePathText {
    NSString *text = [self shellEscapedPath];
    if (text.length == 0) {
        LogHelper(@"cursorpaste: empty file path; cannot type path");
        return;
    }

    NSPasteboard *pb = [NSPasteboard generalPasteboard];
    NSString *priorString = [pb stringForType:NSPasteboardTypeString];
    [pb clearContents];
    [pb setString:text forType:NSPasteboardTypeString];
    LogHelper([NSString stringWithFormat:@"cursorpaste: typing file path %@", text]);
    [self sendPaste];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.05 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [pb clearContents];
        if (priorString.length > 0) {
            [pb setString:priorString forType:NSPasteboardTypeString];
        }
    });
}

- (void)sendLeftClickAt:(NSPoint)point {
    CGEventSourceRef src = CGEventSourceCreate(kCGEventSourceStateCombinedSessionState);
    if (!src) {
        LogHelper(@"cursorpaste: failed to create mouse event source");
        return;
    }

    CGPoint cgPoint = CGPointMake(point.x, point.y);
    CGEventRef down = CGEventCreateMouseEvent(src, kCGEventLeftMouseDown, cgPoint, kCGMouseButtonLeft);
    CGEventRef up = CGEventCreateMouseEvent(src, kCGEventLeftMouseUp, cgPoint, kCGMouseButtonLeft);
    if (down) {
        CGEventPost(kCGHIDEventTap, down);
        CFRelease(down);
    }
    if (up) {
        CGEventPost(kCGHIDEventTap, up);
        CFRelease(up);
    }
    CFRelease(src);
}

- (void)repositionHover {
    if (self.finished) {
        return;
    }
    NSPoint mouse = [NSEvent mouseLocation];
    [self.hoverPanel setFrameOrigin:NSMakePoint(mouse.x + kHoverOffsetX,
                                                mouse.y - self.thumbSize.height - kHoverOffsetX)];
}

- (void)showHover {
    [self.hoverPanel orderFrontRegardless];
}

- (void)hideHover {
    [self.hoverPanel orderOut:nil];
}

- (void)finish:(int)code {
    if (self.finished) return;
    self.finished = YES;

    [self.followTimer invalidate];
    self.followTimer = nil;

    for (id monitor in self.monitors) {
        [NSEvent removeMonitor:monitor];
    }
    [self.monitors removeAllObjects];

    [self hideHover];
    exit(code);
}

@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc < 2) {
            fprintf(stderr, "usage: cursorpaste <image-path>\n");
            return 2;
        }

        NSString *imagePath = [NSString stringWithUTF8String:argv[1]];
        NSImage *image = [[NSImage alloc] initWithContentsOfFile:imagePath];
        if (!image) {
            fprintf(stderr, "cursorpaste: cannot load image %s\n", argv[1]);
            return 2;
        }

        [NSApplication sharedApplication];
        [NSApp setActivationPolicy:NSApplicationActivationPolicyAccessory];

        NSDictionary *options = @{(__bridge NSString *)kAXTrustedCheckOptionPrompt: @YES};
        BOOL trusted = AXIsProcessTrustedWithOptions((__bridge CFDictionaryRef)options);
        if (!trusted) {
            fprintf(stderr, "cursorpaste: Accessibility permission not granted; drag/drop will still work, but click-to-paste needs it.\n");
        } else {
            fprintf(stderr, "cursorpaste: Accessibility permission granted.\n");
        }

        CursorPasteController *controller = [[CursorPasteController alloc] initWithImage:image
                                                                                 fileURL:[NSURL fileURLWithPath:imagePath]
                                                                                 trusted:trusted];
        [controller start];
        [NSApp run];
    }
    return 0;
}

File: cursorpaste.swift
// cursorpaste — float an image on the cursor; click to paste it into whatever
// app you click, Esc to cancel. Used by screenshot-ai's "Auto-delete" path so
// you can drop a screenshot straight into your AI tool, then it self-deletes.
//
// Build: xcrun swiftc -O -o cursorpaste cursorpaste.swift
// Run:   cursorpaste /path/to/image.png    (exit 0 = pasted, 1 = cancelled)
//
// Needs Accessibility permission (to watch for the click/Esc and to send ⌘V).

import Cocoa
import ApplicationServices

let args = CommandLine.arguments
guard args.count >= 2 else {
    FileHandle.standardError.write(Data("usage: cursorpaste <image-path>\n".utf8))
    exit(2)
}
let imagePath = args[1]
guard let image = NSImage(contentsOfFile: imagePath) else {
    FileHandle.standardError.write(Data("cursorpaste: cannot load image \(imagePath)\n".utf8))
    exit(2)
}
let fileURL = URL(fileURLWithPath: imagePath)

let app = NSApplication.shared
app.setActivationPolicy(.accessory) // no Dock icon, don't steal focus

// Prompt for Accessibility on first run. Without it the overlay still follows
// the cursor, but the click/Esc monitors and ⌘V won't work.
let axOpts = [kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: true] as CFDictionary
let trusted = AXIsProcessTrustedWithOptions(axOpts)
if !trusted {
    FileHandle.standardError.write(Data("cursorpaste: needs Accessibility permission — grant it in System Settings → Privacy & Security → Accessibility, then try again.\n".utf8))
}

// --- thumbnail floating panel ---------------------------------------------
let maxSide: CGFloat = 220
let sz = image.size
let scale = min(1, maxSide / max(sz.width, sz.height))
let thumb = NSSize(width: max(40, sz.width * scale), height: max(40, sz.height * scale))

let panel = NSPanel(contentRect: NSRect(origin: .zero, size: thumb),
                    styleMask: [.borderless, .nonactivatingPanel],
                    backing: .buffered, defer: false)
panel.isFloatingPanel = true
panel.level = .screenSaver
panel.backgroundColor = .clear
panel.isOpaque = false
panel.hasShadow = true
panel.ignoresMouseEvents = true // clicks pass through to the app underneath
panel.collectionBehavior = [.canJoinAllSpaces, .fullScreenAuxiliary, .stationary]
panel.alphaValue = 0.92

let iv = NSImageView(frame: NSRect(origin: .zero, size: thumb))
iv.image = image
iv.imageScaling = .scaleProportionallyUpOrDown
iv.wantsLayer = true
iv.layer?.cornerRadius = 8
iv.layer?.masksToBounds = true
iv.layer?.borderWidth = 2
iv.layer?.borderColor = NSColor.white.withAlphaComponent(0.85).cgColor
panel.contentView = iv
panel.orderFrontRegardless()

// Follow the cursor (offset down-right so the cursor stays visible).
func reposition() {
    let m = NSEvent.mouseLocation
    panel.setFrameOrigin(NSPoint(x: m.x + 14, y: m.y - thumb.height - 14))
}
reposition()
let follow = Timer.scheduledTimer(withTimeInterval: 1.0 / 120.0, repeats: true) { _ in reposition() }
RunLoop.main.add(follow, forMode: .common)

// --- paste + exit ----------------------------------------------------------
func copyToPasteboard() {
    let pb = NSPasteboard.general
    pb.clearContents()
    let item = NSPasteboardItem()
    if let tiff = image.tiffRepresentation,
       let rep = NSBitmapImageRep(data: tiff),
       let png = rep.representation(using: .png, properties: [:]) {
        item.setData(png, forType: .png)
        item.setData(tiff, forType: .tiff)
    }
    pb.writeObjects([item, fileURL as NSURL]) // image data + the file itself
}

func sendPaste() {
    let src = CGEventSource(stateID: .combinedSessionState)
    let vKey: CGKeyCode = 0x09 // 'v'
    let down = CGEvent(keyboardEventSource: src, virtualKey: vKey, keyDown: true)
    down?.flags = .maskCommand
    let up = CGEvent(keyboardEventSource: src, virtualKey: vKey, keyDown: false)
    up?.flags = .maskCommand
    down?.post(tap: .cghidEventTap)
    up?.post(tap: .cghidEventTap)
}

var finished = false
func finish(_ code: Int32) {
    if finished { return }
    finished = true
    follow.invalidate()
    panel.orderOut(nil)
    exit(code)
}

// Esc cancels.
NSEvent.addGlobalMonitorForEvents(matching: .keyDown) { ev in
    if ev.keyCode == 53 { finish(1) } // 53 = Escape
}

// A click drops it: the click focuses the target field, then we paste into it.
NSEvent.addGlobalMonitorForEvents(matching: .leftMouseUp) { _ in
    copyToPasteboard()
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.08) {
        sendPaste()
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) { finish(0) }
    }
}

app.run()
